### PR TITLE
Migrate `pkg/controller/bootstrap` to contextual logging

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Migrate `pkg/controller/bootstrap` to contextual logging

#### Which issue(s) this PR fixes:

Part of  https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note
Migrated the bootstrap signer controller and the token cleaner controller (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

### CODE REVIEW

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/bootstrapsigner.go:200]:
```suggestion
		logger.V(3).Info("No KubeConfigKey in ConfigMap", "kubeConfigKey", bootstrapapi.KubeConfigKey, "configMap", klog.KRef(origCM.Namespace, origCM.Name))
```

File: pkg/controller/bootstrap/tokencleaner.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/tokencleaner.go:171]:
```suggestion
		logger.V(3).Info("Secret has been deleted", "secret", key)
```

File: pkg/controller/bootstrap/util.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/util.go:23]: Maybe  "context"?

File: pkg/controller/bootstrap/util.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/util.go:36]:
```suggestion
		logger.V(3).Info("Invalid secret name. Must be the right form", "secretName", secret.Name, "form", form)
```

File: pkg/controller/bootstrap/util.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/util.go:47]:
```suggestion
		logger.V(3).Info("Token ID doesn't match secret name", "tokenID", tokenID, "secretName", secret.Name)
```

File: pkg/controller/bootstrap/util.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util.go:23]: Oops! the IDE automatically imported the incorrect context, I didn't notice. Thanks!

File: pkg/controller/bootstrap/util.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util.go:36]: My bad. I left it out to delete it. Thanks!

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: done.

File: pkg/controller/bootstrap/tokencleaner.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/tokencleaner.go]: Updated.

File: pkg/controller/bootstrap/util.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util.go]: done.

File: pkg/controller/bootstrap/util_test.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/util_test.go:21]: Maybe it's better to adjust the format

File: pkg/controller/bootstrap/util.go
Comment (yangjunmyfm192085) [inline pkg/controller/bootstrap/util.go:27]: Why do we need to modify this?

File: pkg/controller/bootstrap/util.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util.go]: done.

File: pkg/controller/bootstrap/util_test.go
Comment (logicalhan) [inline pkg/controller/bootstrap/util_test.go:28]:
```suggestion
	"k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	bootstrapapi "k8s.io/cluster-bootstrap/token/api"
	"k8s.io/klog/v2/ktesting"
```

**Review by logicalhan — COMMENTED** [review]
/lgtm

File: pkg/controller/bootstrap/util_test.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util_test.go]: Have updated it.

File: pkg/controller/bootstrap/util_test.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util_test.go]: done

File: pkg/controller/bootstrap/util_test.go
Comment (pohly) [inline pkg/controller/bootstrap/util_test.go:86]: `WithCancel` doesn't seem to be needed here.

Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util_test.go]: Done.

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/bootstrap/tokencleaner.go
Comment (pohly) [inline pkg/controller/bootstrap/tokencleaner.go:190]:
```suggestion
		logger.V(3).Info("Deleting expired secret", "secret", klog.KObj((secret))
```

File: pkg/controller/bootstrap/util_test.go
Comment (pohly) [inline pkg/controller/bootstrap/util_test.go:102]: You can remove the `%s:` and `tc.description`. Because these are now sub-tests, the test failure will automatically be associated with `tc.description` because its the name of the sub-test.

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/bootstrap/tokencleaner.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/tokencleaner.go]: done.

File: pkg/controller/bootstrap/util_test.go
Comment (mengjiao-liu) [reply pkg/controller/bootstrap/util_test.go]: Yes, `%s:` and `tc.description` should be deleted.

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by mengjiao-liu** [comment]
pull-kubernetes-unit error info：
```
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
    output.go:54: 
        	Error Trace:	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/internal/unittests/cleanup/output.go:54
        	Error:      	Not equal: 
        	            	expected: "[BeforeEach] e2e\n  cleanup_test.go:53\nINFO: before\n[BeforeEach] e2e\n  set up framework | framework.go:xxx\nSTEP: Creating a kubernetes client\nINFO: >>> kubeConfig: yyy/kube.config\nSTEP: Building a namespace api object, basename test-namespace\nINFO: Skipping waiting for service account\n[BeforeEach] e2e\n  cleanup_test.go:95\nINFO: extension before\n[BeforeEach] e2e\n  cleanup_test.go:61\nINFO: before #1\n[BeforeEach] e2e\n  cleanup_test.go:65\nINFO: before #2\n[It] works\n  cleanup_test.go:80\n[AfterEach] e2e\n  cleanup_test.go:96\nINFO: extension after\n[AfterEach] e2e\n  cleanup_test.go:69\nINFO: after #1\n[AfterEach] e2e\n  cleanup_test.go:76\nINFO: after #2\n[DeferCleanup (Each)] e2e\n  cleanup_test.go:85\nINFO: cleanup first\n[DeferCleanup (Each)] e2e\n  cleanup_test.go:82\nINFO: cleanup last\n[DeferCleanup (Each)] e2e\n  dump namespaces | framework.go:xxx\n[DeferCleanup (Each)] e2e\n  tear down framework | framework.go:xxx\nSTEP: Destroying namespace \"test-namespace-zzz\" for this suite.\n"
        	            	actual  : "[BeforeEach] e2e\n  cleanup_test.go:53\nINFO: before\n[BeforeEach] e2e\n  set up framework | framework.go:xxx\nSTEP: Creating a kubernetes client\nINFO: >>> kubeConfig: yyy/kube.config\nSTEP: Building a namespace api object, basename test-namespace 12/13/22 11:43:53\nINFO: Skipping waiting for service account\n[BeforeEach] e2e\n  cleanup_test.go:95\nINFO: extension before\n[BeforeEach] e2e\n  cleanup_test.go:61\nINFO: before #1\n[BeforeEach] e2e\n  cleanup_test.go:65\nINFO: before #2\n[It] works\n  cleanup_test.go:80\n[AfterEach] e2e\n  cleanup_test.go:96\nINFO: extension after\n[AfterEach] e2e\n  cleanup_test.go:69\nINFO: after #1\n[AfterEach] e2e\n  cleanup_test.go:76\nINFO: after #2\n[DeferCleanup (Each)] e2e\n  cleanup_test.go:85\nINFO: cleanup first\n[DeferCleanup (Each)] e2e\n  cleanup_test.go:82\nINFO: cleanup last\n[DeferCleanup (Each)] e2e\n  dump namespaces | framework.go:xxx\n[DeferCleanup (Each)] e2e\n  tear down framework | framework.go:xxx\nSTEP: Destroying namespace \"test-namespace-zzz\" for this suite.\n"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -7,3 +7,3 @@
        	            	 INFO: >>> kubeConfig: yyy/kube.config
        	            	-STEP: Building a namespace api object, basename test-namespace
        	            	+STEP: Building a namespace api object, basename test-namespace 12/13/22 11:43:53
        	            	 INFO: Skipping waiting for service account
        	Test:       	TestCleanup
        	Messages:   	output from test #0 (e2e works)
    klog.go:876: INFO Shutting down kubernetes service endpoint reconciler
```
source code file：test/e2e/framework/internal/unittests/cleanup/cleanup_test.go

It seems that the error is that the expected content does not match the actual content.  The actual content is different with timestamps
https://github.com/kubernetes/kubernetes/blob/10376ad1fae2f23d8de0e2b9cf4c24a7f0edc547/test/e2e/framework/internal/unittests/cleanup/cleanup_test.go#L109
```
        	            	-STEP: Building a namespace api object, basename test-namespace
        	            	+STEP: Building a namespace api object, basename test-namespace 12/13/22 11:43:53
```

It seems that it has nothing to do with this PR, it may be caused by other PRs, I need to look at it carefully recently.

**Comment by pohly** [comment]
You hit a full second with that output, which made this regexp not match it:
```
// timeSuffix matches "09/06/22 15:36:43.445" as printed by Ginkgo v2 for log output.
var timeSuffix = regexp.MustCompile(`(?m)[[:space:]][[:digit:]]{2}/[[:digit:]]{2}/[[:digit:]]{2} [[:digit:]]{2}:[[:digit:]]{2}:[[:digit:]]{2}\.[[:digit:]]{1,3}$`)
```

I'll fix it.

/retest

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (atiratree) [inline cmd/kube-controller-manager/app/bootstrap.go:30]: 1. Do we care that there is a one character difference `-` from the [controller name](https://github.com/kubernetes/kubernetes/blob/1e3946ce9dcb745771b03845d4afd97caa7c7737/cmd/kube-controller-manager/app/controllermanager.go#L455) `bootstrapsigner`? From consistency perspective, people could be potentially used to the controller name from running the KCM (`--controllers`). Not sure if that is an actual problem though.

2. There is also an inconsistent naming across controllers: just by picking a few
  - deployment
  - cronjob-controller
  - NodeIpamController

We should choose a consistent format/naming across all of the controllers.

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (atiratree) [inline cmd/kube-controller-manager/app/bootstrap.go:45]: maybe we could consider having a function for this, if this is a common pattern?

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (atiratree) [inline pkg/controller/bootstrap/bootstrapsigner.go:158]: I suppose the function above will consumes the context, once the client-go refactoring goes through? 

I don't know what is easier from refactoring complexity POV to do/merge first. But this will result in some of the statement having the `bootstrap-signer` prefix and some statements that are printed via client-go without the prefix even though they belong to the same controller. 

Are we expecting to merge the client-go refactoring in the same release (1.27) ?

Comment (atiratree) [reply pkg/controller/bootstrap/bootstrapsigner.go:158]: the same goes for apimachinery and [ErrorHandlers](https://github.com/kubernetes/kubernetes/blob/1e3946ce9dcb745771b03845d4afd97caa7c7737/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L87) for example. And other dependencies.

File: pkg/controller/bootstrap/tokencleaner.go
Comment (atiratree) [inline pkg/controller/bootstrap/tokencleaner.go:158]: noobie question: why do we not create the logger per component and obtain it from `TokenCleaner` struct? Is it because there should be only a single place which should the logging honor ( `ctx` variable in this case) as outlined in the KEP to prevent ambiguity which one to use?

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (pohly) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: Consistency is important. If there are already established names, then please point them out as part of your review and use `/lgtm cancel` if necessary to track PRs that aren't ready yet.

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (pohly) [reply cmd/kube-controller-manager/app/bootstrap.go:45]: It is common, not just in kube-controller-manager, but also elsewhere. However, the only purpose of a helper function would be to avoid some typing. I prefer to see in the code what's going on instead of having to look up a helper function with a one-line body elsewhere - might just be my personal preference, though.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (pohly) [reply pkg/controller/bootstrap/bootstrapsigner.go:158]: Which client-go refactoring? Is that some specific effort or do you mean the migration of client-go to contextual logging?

I started that migration, but hadn't planned to submit it for 1.27 because some aspects of it a bit tricky and may require API changes. On the other hand you are right, unless all dependencies that log something also get migrated, a binary will have inconsistent logging. We are still in the alpha phase of contextual logging and this will not be noticeable unless someone explicitly flips the "ContextualLogging" feature gate, so that's okay?

File: pkg/controller/bootstrap/tokencleaner.go
Comment (pohly) [reply pkg/controller/bootstrap/tokencleaner.go:158]: Yes, that's exactly the problem.

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (atiratree) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: for the 1. point I think both sides make sense. The logs will get much more exposure and could make sense to make them readable. I will let @soltysh to give his view on this.

2. so in that case I would suggest we use the following format everywhere
- deployment-controller
- cronjob-controller
- nodeipam-controller or node-ipam-controller

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (atiratree) [reply pkg/controller/bootstrap/bootstrapsigner.go:158]: > Which client-go refactoring? Is that some specific effort or do you mean the migration of client-go to contextual logging?

I meant the contextual logging

> I started that migration, but hadn't planned to submit it for 1.27 because some aspects of it a bit tricky and may require API changes. On the other hand you are right, unless all dependencies that log something also get migrated, a binary will have inconsistent logging. We are still in the alpha phase of contextual logging and this will not be noticeable unless someone explicitly flips the "ContextualLogging" feature gate, so that's okay?

Thanks for explanation. I suppose that is fine if we are planning to wait with switching the gate on by default, until we can ensure a good experience.

File: pkg/controller/bootstrap/tokencleaner.go
Comment (atiratree) [reply pkg/controller/bootstrap/tokencleaner.go:158]: ok, thanks!

**Comment by atiratree** [comment]
lgtm, apart from the naming convention which is still up for a discussion: https://github.com/kubernetes/kubernetes/pull/113464#discussion_r1065155841

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (atiratree) [reply cmd/kube-controller-manager/app/bootstrap.go]: initiated a conversation in k8s dev: https://groups.google.com/u/1/a/kubernetes.io/g/dev/c/I9l8SSM599s

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (liggitt) [inline cmd/kube-controller-manager/app/bootstrap.go:30]: would it make more sense for the caller to wrap the context to get contextual logging before calling the start___Controller method, so the thing setting up all the controller loops could guarantee this was done consistently, and could do so using the identifier already associated with each controller loop (like the admin-facing initializer name)?

Comment (pohly) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: That makes sense. I suppose that would be in [StartControllers](https://github.com/kubernetes/kubernetes/blob/eb4e2a2225bd534aa7d765510fbf9e441d498c5e/cmd/kube-controller-manager/app/controllermanager.go#L566-L567).

Is the initializer name the string set in  [NewControllerInitializers](https://github.com/kubernetes/kubernetes/blob/eb4e2a2225bd534aa7d765510fbf9e441d498c5e/cmd/kube-controller-manager/app/controllermanager.go#LL423C5-L423C31)?

@atiratree: my preference would be to not ask all the different contributors to revise their PRs, but instead merge them when the rest of the changes are okay. Then once they are merged, I can prepare a PR which removes the individual names from all controller and instead does it in StartControllers. It will be less work overall.

Comment (atiratree) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: that is a good idea - to move the responsibility for consistency outside

@pohly yup, doing it in one sweep sounds good (as we are still alpha)

Comment (atiratree) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: and yes we would use the initializer names set (after we make it consistent)

Comment (liggitt) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: > That makes sense. I suppose that would be in [StartControllers](https://github.com/kubernetes/kubernetes/blob/eb4e2a2225bd534aa7d765510fbf9e441d498c5e/cmd/kube-controller-manager/app/controllermanager.go#L566-L567).

Exactly

> Is the initializer name the string set in [NewControllerInitializers](https://github.com/kubernetes/kubernetes/blob/eb4e2a2225bd534aa7d765510fbf9e441d498c5e/cmd/kube-controller-manager/app/controllermanager.go#LL423C5-L423C31)?

Yes

> @atiratree: my preference would be to not ask all the different contributors to revise their PRs, but instead merge them when the rest of the changes are okay. Then once they are merged, I can prepare a PR which removes the individual names from all controller and instead does it in StartControllers. It will be less work overall.

If we haven't merged the changes to put context logging construction inside controllers yet, it seems questionable to proceed adding that to all the controllers only to then remove it.

Comment (pohly) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: I avoids extra roundtrips with contributors who may or may not react to further change requests in a timely manner.  It's of course cleaner to not add these lines in the first place.

Comment (pohly) [reply cmd/kube-controller-manager/app/bootstrap.go:30]: Discussed this with @soltysh and we agreed to move ahead with these PRs as they are and finish up later.

/lgtm

File: cmd/kube-controller-manager/app/bootstrap.go
Comment (atiratree) [reply cmd/kube-controller-manager/app/bootstrap.go]: for posterity, implementing aliases in https://github.com/kubernetes/kubernetes/pull/115813

**Comment by pohly** [comment]
/assign @soltysh 

For approval.

**Comment by mengjiao-liu** [comment]
ping @soltysh

**Review by soltysh — APPROVED** [review]
/approve

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (pacoxu) [inline pkg/controller/bootstrap/bootstrapsigner.go:200]: /hold
verify ci failed.

Comment (mengjiao-liu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: It looks like the `logtools` validation tool has been updated. I  have updated the PR for validation.

**Pull request merged**
